### ISSUE TITLE
Installation issue: libdwarf

### ISSUE THREAD

**Issue by jacobmerson** [spack/spack#15735]
### Spack version

```console
$ spack --version
0.14.1-415-cbcac7225
```

### Steps to reproduce the issue

```console
$ spack install hpctoolkit+papi+cuda %gcc@7.3.0 ^binutils@2.32%gcc@7.3.0 ^dyninst@10.1.0%gcc@7.3.0 
Error: ProcessError: Command exited with status 2:                                               
    'make' 'ARFLAGS=rcs'
```
If I try to install libdwarf directly, I get other dependencies failing, and the end result I care about is installing hpctoolkit which is why I list that as the spec.

### Platform and user environment

```commandline
$ uname -a 
Linux blockade 3.10.0-1062.1.2.el7.x86_64 #1 SMP Mon Sep 16 14:19:51 EDT 2019 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:	Red Hat Enterprise Linux Server release 7.7 (Maipo)
``` 

### Additional information
```console
$ spack spec --install-status <spec>
Input spec
--------------------------------
 -   hpctoolkit%gcc@7.3.0+cuda+papi
 -       ^binutils@2.32%gcc@7.3.0
 -       ^dyninst@10.1.0%gcc@7.3.0

Concretized
--------------------------------
 -   hpctoolkit@2020.03.01%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" ~all-static~bgq~cray+cuda~mpi+papi arch=linux-rhel7-x86_64
[+]      ^binutils@2.32%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +gold~headers+libiberty~nls~plugins arch=linux-rhel7-x86_64
[^]          ^zlib@1.2.11%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +optimize+pic+shared arch=linux-rhel7-x86_64
[+]      ^boost@1.72.0%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +atomic+chrono~clanglibcpp~container~context~coroutine cxxstd=98 +date_time~debug+exception~fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout visibility=global +wave arch=linux-rhel7-x86_64
[^]          ^bzip2@system%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +shared arch=linux-rhel7-x86_64
[+]      ^cuda@10.1%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2"  arch=linux-rhel7-x86_64
[+]      ^dyninst@10.1.0%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2"  build_type=RelWithDebInfo +openmp~stat_dysect~static arch=linux-rhel7-x86_64
[+]          ^cmake@3.16.5%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" ~doc+ncurses+openssl+ownlibs~qt arch=linux-rhel7-x86_64
[+]              ^ncurses@6.2%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" ~symlinks+termlib arch=linux-rhel7-x86_64
[^]                  ^pkgconf@1.6.3%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2"  arch=linux-rhel7-x86_64
[+]              ^openssl@1.1.1e%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +systemcerts arch=linux-rhel7-x86_64
[+]                  ^perl@5.16.3%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +cpanm patches=0eac10ed90aeb0459ad8851f88081d439a4e41978e586ec743069e8b059370ac +shared+threads arch=linux-rhel7-x86_64
[+]          ^elfutils@0.178%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +bzip2~nls+xz arch=linux-rhel7-x86_64
[^]              ^m4@1.4.16%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +sigsegv arch=linux-rhel7-x86_64
[^]              ^xz@system%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2"  arch=linux-rhel7-x86_64
[+]          ^intel-tbb@2020.1%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2"  cxxstd=default patches=62ba015ebd1819c45bef47411540b789b493e31ca668c4ff4cb2afcbc306b476,ce1fb16fb932ce86a82ca87cf0431d1a8c83652af9f552b264213b2ff2945d73 +shared+tm arch=linux-rhel7-x86_64
[+]          ^libiberty@2.33.1%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +pic arch=linux-rhel7-x86_64
[^]      ^intel-xed@2019.03.01%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" ~debug arch=linux-rhel7-x86_64
[+]          ^python@2.7.5%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4~uuid+zlib arch=linux-rhel7-x86_64
 -       ^libdwarf@20180129%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2"  arch=linux-rhel7-x86_64
[^]      ^libmonitor@2018.07.18%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" ~bgq+hpctoolkit~krellpatch arch=linux-rhel7-x86_64
[+]      ^libunwind@1.4-rc1%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +xz~zlib arch=linux-rhel7-x86_64
[+]      ^mbedtls@2.16.1%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2"  build_type=Release +pic arch=linux-rhel7-x86_64
[+]      ^papi@6.0.0%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2" +example~infiniband~lmsensors~powercap~rapl arch=linux-rhel7-x86_64
[^]      ^xerces-c@3.2.2%gcc@7.3.0 cflags="-mtune=native -march=core2" cxxflags="-mtune=native -march=core2" fflags="-mtune=native -march=core2"  cxxstd=default transcoder=iconv arch=linux-rhel7-x86_64
```
to show people whether Spack installed a faulty software or if it was not able to
build it at all. 

If your build didn't make it past the configure stage, Spack as also commands to parse 
logs and report error and warning messages:
```console
$ spack log-parse --show=errors,warnings build-out-errors.txt
5 errors
0 warnings
     110    /lore/mersoj/spack/lib/spack/env/gcc/gcc   -g -O2 -I. -I. -fPIC       -c -o dwarf_arange
            .o dwarf_arange.c
     111    /lore/mersoj/spack/lib/spack/env/gcc/gcc   -g -O2 -I. -I. -fPIC       -c -o dwarf_die_de
            liv.o dwarf_die_deliv.c
     112    dwarf_die_deliv.c: In function '_dwarf_make_CU_Context':
  >> 113    dwarf_die_deliv.c:686:13: error: 'DW_SECT_ABBREV' undeclared (first use in this function
            ); did you mean 'DW_DLA_ABBREV'?
     114                 DW_SECT_ABBREV, &absize);
     115                 ^~~~~~~~~~~~~~
     116                 DW_DLA_ABBREV
     117    dwarf_die_deliv.c:686:13: note: each undeclared identifier is reported only once for eac
            h function it appears in
     118    dwarf_die_deliv.c: In function 'find_context_base_fields':
  >> 119    dwarf_die_deliv.c:925:32: error: 'DW_AT_dwo_id' undeclared (first use in this function);
             did you mean 'DW_ATE_void'?
     120                     if (attrnum == DW_AT_dwo_id ||
     121                                    ^~~~~~~~~~~~
     122                                    DW_ATE_void
     123    dwarf_die_deliv.c: In function 'dwarf_die_from_hash_signature':
  >> 124    dwarf_die_deliv.c:1346:22: error: 'DW_SECT_INFO' undeclared (first use in this function)
            ; did you mean 'DT_SYMINFO'?
     125                 is_info2?DW_SECT_INFO:DW_SECT_TYPES,
     126                          ^~~~~~~~~~~~
     127                          DT_SYMINFO
  >> 128    dwarf_die_deliv.c:1346:35: error: 'DW_SECT_TYPES' undeclared (first use in this function
            ); did you mean 'DW_SECT_INFO'?
     129                 is_info2?DW_SECT_INFO:DW_SECT_TYPES,
     130                                       ^~~~~~~~~~~~~
     131                                       DW_SECT_INFO
  >> 132    make: *** [dwarf_die_deliv.o] Error 1
```
### General information

- [x ] I have run `spack --version` and reported the version of Spack
- [x ] I have searched the issues of this repo and believe this is not a duplicate

**Comment by mwkrentel** [comment]
@jacobmerson I have a nearly identical Centos 7.7 system and it works for me.

I can think of two main possibilities.

1. You have something messed up (inconsistent) in your environment.
I've seen this cause all sorts of strange errors where other users
report "works for me."

2. There's something missing on your system that libdwarf implicitly
needs that should belong in the spack recipe but isn't.
That would say the libdwarf recipe is incomplete, but it's a bit
odd that no one has noticed before.

Do you have multiple copies of spack checked out?

Any chance you sourced the setup scripts from an old spack that you're
using in this build (and thus compiler flags point to things that
don't exist).  This would be consistent with the error messages.

If it's not something simple like above, then repeat all the steps
from scratch.  Use a fresh xterm, fresh spack clone, make sure there
is nothing in your environment related to spack, don't source the
setup scripts, just run spack directly from the bin directory, etc.
Make sure there's nothing pointing to an old install.  Run:

```
spack install -v libdwarf %gcc@7.3.0 ^elfutils +bzip2+xz~nls  2>&1 | tee out.err
```

If that fails, then send us the out.err file, or at least the part related
to libdwarf.

Also, RH 7.x comes with gcc 4.8, so I assume that 7.3.0 is from a
module.  What does `spack compiler info gcc` report?
I want to check that the `modules` field is set.

Btw, you don't need to specify `%gcc@7.3.0` multiple times.  Also,
dyninst 10.1.0 is the default, so this should be equivalent.

```
spack spec -I hpctoolkit+cuda+papi %gcc@7.3.0 ^binutils@2.32  
```

**Comment by jacobmerson** [comment]
Thanks for getting back to me. If there is an inconsistency, I'm wondering if it is cropping up because I was trying to "chain" my spack installation against a system wide version to avoid rebuilding everything which is globally installed.

As suggested, I am starting over with a clean spack installation. Do you tend to put anything in your packages.yaml file? i.e. python,perl,xz, or let spack build everything from scratch?

**Comment by jacobmerson** [comment]
If I start with a completely clean shell, don't include the upstream spack installaiton, and build everything from scratch using spack (i.e. no system packages besides compilers), everything builds properly.

I can experiment to see if I am able to get the upstream packages working properly, but this issue can be closed.

@mwkrentel thanks again for your suggestions.

**Comment by mwkrentel** [comment]
@jacobmerson  I'm not sure what you mean by "chaining" and "upstream spack
installation".  These must be features that I don't use.
Do you have a reference in spack read the docs?

As for packages.yaml, I normally use the system (/usr) version for
cmake, perl and python, assuming the versions are recent enough.
For RH 7.7 they should be.  But this is only for build tools.
Anything that has libraries that we use at runtime, I always build
from scratch.

**Comment by jacobmerson** [comment]
Here is the documentation on "chaining". https://spack.readthedocs.io/en/latest/chain.html

I'm not positive if this was my issue or not, but now that I have something working I can work backwards to figure out what breaks the install...

**Comment by jacobmerson** [comment]
Weirdly enough the issues I was having seemed to come from running spack inside of tmux. I'll probably open a new issue to see if anyone else can replicate.
